mirage/awa_mirage.ml

```diff
@@ -214,46 +214,70 @@ module Make (F : Mirage_flow.S) (T : Mirage_time.S) (M : Mirage_clock.MCLOCK) =
   let lookup_channel t id =
     List.find_opt (fun c -> id = c.id) t.channels
 
-  let rec nexus t fd server input_buffer =
+  let rec nexus t fd server input_buffer pending_promises =
     wrapr (Awa.Server.pop_msg2 server input_buffer)
     >>= fun (server, msg, input_buffer) ->
     match msg with
     | None -> (* No SSH msg *)
       Lwt.catch
         (fun () ->
           let timeout = T.sleep_ns 2000000000L >>= fun () -> Lwt.return Rekey in
-          Lwt.pick [ Lwt_mvar.take t.nexus_mbox ;
-                      net_read fd ;
-                      timeout ])
+          (* We will listen from two incomming messages sources, from the net interface with
+           * 'net_read', and from the ssh server with 'Lwt_mvar.take'. To let the promises
+           * to be resolved, we use Lwt.choose to not add another of these until we know
+           * that it was fullfiled.
+           *)
+          Lwt.nchoose_split (List.append pending_promises [ timeout ])
+        )
       (function exn -> Lwt.fail exn)
-      >>= fun nexus_msg ->
-      (match nexus_msg with
-        | Rekey ->
-          (match Awa.Server.maybe_rekey server (now ()) with
-          | None -> nexus t fd server input_buffer
-          | Some (server, kexinit) ->
-            send_msg fd server kexinit
-            >>= fun server ->
-            nexus t fd server input_buffer)
-       | Net_eof ->
-         Lwt.return t
-       | Net_io buf -> nexus t fd server (Awa.Util.cs_join input_buffer buf)
-       | Sshout (id, buf) | Ssherr (id, buf) ->
-         wrapr (Awa.Server.output_channel_data server id buf)
-         >>= fun (server, msgs) ->
-         send_msgs fd server msgs >>= fun server ->
-         nexus t fd server input_buffer)
+      >>= fun (nexus_msg_fullfiled, pending_promises) ->
+      (* We need to keep track of the "not fullfiled" promises and only Lwt.nchoose_split
+       * allows us to have this information. This function also gives us a list of
+       * "already fullfiled" promises. Here we consume this list and add the relevant new
+       * promises to watch.
+       *)
+      let rec loop t fd server input_buffer fullfiled_promises pending_promises =
+        if (List.length fullfiled_promises == 0) then
```

Comment:
You can use a match instead:
```OCaml
match fulfilled_promises with
| [] -> nexus ... pending_promises
| Rekey :: remaining_fulfilled_promises -> ...
| Net_eof :: _ -> ...
| Net_io buf :: remaining_fulfilled_promises -> ...
```

Reply:
Thanks for the comments, this will be mush easier to read